### ISSUE TITLE
Unable to connect via proxy to port 80 with method CONNECT

### ISSUE THREAD

**Issue by ingoj** [friendica/friendica#4734]
Repost of https://nerdica.net/display/a85d7459785ac29244dc380734848258

When configuring a Proxy URL in Friendicas Site config I have the problem that CONNECT method on port 80 is being refused by the proxy on the same host. Connections with CONNECT method via port 443 is working just fine.

When using tinyproxy I get this error:

CONNECT Apr 02 22:19:39 [23281]: Connect (file descriptor 6): 192.168.254.2 [192.168.254.2]
CONNECT Apr 02 22:19:39 [23281]: Request (file descriptor 6): CONNECT api.openweathermap.org:80 HTTP/1.1
INFO Apr 02 22:19:39 [23281]: Refused CONNECT method on port 80
INFO Apr 02 22:19:39 [23281]: no entity

When using Squid3 I get this error:

1522699829.434 0 192.168.254.2 TCP_DENIED/403 3966 CONNECT api.openweathermap.org:80 - HIER_NONE/- text/html

From what I found out searching for a solution on the web "CONNECT" is only being used for HTTPS connection to tell the proxy to connect to that specific host. This would render connections with "CONNECT" method via port 80 illegal.

When this is true, this would be a bug in Friendica, I guess. Can someone confirm/deny?

Running Friendica 3.6 on Debian Stable.

**Comment by MrPetovan** [comment]
I'm sorry, I'm not familiar with proxies.

**Comment by ingoj** [comment]
Well, maybe some resources to read: 
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/CONNECT
- https://tools.ietf.org/html/rfc7231#section-4.3.6

Please note that even the RFC has listed port 80: 
     CONNECT server.example.com:80 HTTP/1.1
     Host: server.example.com:80
     Proxy-Authorization: basic aGVsbG86d29ybGQ=

But it also states: 
   Proxies that
   support CONNECT SHOULD restrict its use to a limited set of known
   ports or a configurable whitelist of safe request targets.

I assume that squid and tinyproxy are interpreting this but their default config to only allow method CONNECT to SSL ports. I've tried to set the ACLs within squid to also allow Safe_ports (which includes port 80), but this didn't work either. 

https://en.wikipedia.org/wiki/HTTP_tunnel#HTTP_CONNECT_method says: 
"This mechanism is how a client behind an HTTP proxy can access websites using SSL or TLS (i.e. HTTPS). Proxy servers may also limit connections by only allowing connections to the default HTTPS port 443, whitelisting hosts, or blocking traffic which doesn't appear to be SSL."

And: https://en.wikipedia.org/wiki/HTTP_tunnel#HTTP_tunneling_without_using_CONNECT
"Since all traffic is encapsulated inside normal GET and POST requests and responses, this approach works through most proxies and firewalls."

My assumption is: 
CONNECT is used to establish end-to-end tunnel to port 443 (HTTPS). Port 80 is not secure and therefor the proxy is refusing to establish an end-to-end tunnel, because one end is not secure. Therefor requests to port 80 should be encapsulated in GET or POST requests instead of CONNECT. 

But I'm neither an expert in proxies nor in reading/interpreting RFCs.

**Comment by MrPetovan** [comment]
Thanks for the elaboration, although I'm not sure either where in the Friendica code we are sending CONNECT requests. Hopefully your explanation will be able to help someone else troubleshoot the issue.

**Comment by ingoj** [comment]
I've solved this issue somehow. 

For some reasons (coming from a very old Debian version) I've had squid and squid3 packages installed. Squid3 package is now just a transitional package. However, the whole config/directory setup was mixed up between squid and squid3 package. Somehow in daily operation this didn't have any impact in normal operation as a proxy. In this case it did for Friendica. After I removed squid3 transitional package and fixed some config paths and such, Friendica is now able to CONNECT to port 80 as well: 

1522841243.401     11 192.168.254.2 TCP_TUNNEL/200 1118 CONNECT api.openweathermap.org:80 - HIER_DIRECT/95.85.63.65 -

Only problem I now see, is, that it is using a TUNNEL and apparently TUNNEL connections are not cacheable as it seems, because every such connection is a direct connection (HIER_DIRECT). Usually I would expect such a log entry for a cache hit: 

TCP_HIT/200 49572 GET

At least the CONNECT to port 80 issue is solved and can be closed. However, I'm not sure if Friendica is doing the right thing when using a proxy. The purpose of a proxy is also to cache content. With the current setup, cacheing will never happen, I think. For this Friendica would need to use method GET instead of CONNECT, I believe. This could then save some bandwidth and traffic, and reduce page load times, I believe.
